otherlibs/fiber/pool.ml

```diff
@@ -20,26 +16,58 @@ let running t k =
   | Open -> k true
   | Closed -> k false
 
-let create () = { mvar = Mvar.create (); status = Open }
+let create () = { tasks = Queue.create (); runner = None; status = Open }
 
 let task t ~f k =
   match t.status with
   | Closed ->
     Code_error.raise "pool is closed. new tasks may not be submitted" []
-  | Open -> Mvar.write t.mvar (Task f) k
-
-let stream t =
-  Stream.In.create (fun () ->
-      let+ next = Mvar.read t.mvar in
-      match next with
-      | Done -> None
-      | Task task -> Some task)
+  | Open -> (
+    match t.runner with
+    | None ->
+      Queue.push t.tasks f;
+      k ()
+    | Some r ->
+      assert (Queue.is_empty t.tasks);
+      t.runner <- None;
+      Queue.push t.tasks f;
+      resume r () k)
 
 let stop t k =
   match t.status with
   | Closed -> k ()
-  | Open ->
+  | Open -> (
     t.status <- Closed;
-    Mvar.write t.mvar Done k
+    match t.runner with
+    | None -> k ()
+    | Some r ->
+      t.runner <- None;
+      resume r () k)
 
-let run t = stream t |> Stream.In.parallel_iter ~f:(fun task -> task ())
+let run t k =
+  match t.runner with
+  | Some _ ->
+    Code_error.raise "Fiber.Pool.run: concurent calls to run aren't allowed" []
+  | None ->
+    let n = ref 1 in
+    let done_fiber () =
+      decr n;
+      if !n = 0 then k () else end_of_fiber
+    in
+    let rec read t =
+      match Queue.pop t.tasks with
+      | None -> on_finish t
+      | Some v ->
+        incr n;
+        fork (fun () -> v () done_fiber) read_delayed
+    and read_delayed () = read t
+    and suspend_k k =
+      assert (t.runner = None);
+      assert (Queue.is_empty t.tasks);
```

Comment:
It's not obvious that this `assert` can't fail, since this seems to depend on the semantics of `suspend`. It this statically checkable, i.e. could we turn the runner type into a GADT that tracks emptiness to eliminate this assert? (It might be not worth the trouble, of course.) 

I think I'm mostly missing some top-level explanation for how/why this code works.

Reply:
> It this statically checkable, i.e. could we turn the runner type into a GADT that tracks emptiness to eliminate this assert? (It might be not worth the trouble, of course.)

It might be possible if we bake in `Pool` into `effect` like we do with `Ivar`, but I don't think it's possible here. The problem is that we only want to allocate the `suspend_k` closure once, so it must acquire `t` from the environment. Which means that it will only always get the unrefined `'a t`. Or at least I'm not sure how to do it.

Reply:
Come to think of it, baking `Pool` into `effect` will make it even faster. So it might be worth doing one day anyway. I'll abstain for now.

Reply:
Sounds good. And do mention this idea in the comment too :)